### ISSUE TITLE
Github input for both ssh and tokens

### ISSUE THREAD

**Issue by jkarni** [NixOS/nix#10056]
**Is your feature request related to a problem? Please describe.**
There are several different input types - `git+https`, `github`, etc - and between those you can handle private flake inputs from git/github. A problem I've run into repeatedly, however, is that I have to choose between using SSH/HTTPS or tokens, when in practice in some scenarios (e.g., local development) I want SSH, and in some (e.g., CI) I want tokens.

**Describe the solution you'd like**
I'd like a fetcher that tries both SSH and tokens, but without changing the flake lock file to make it incompatible with the other choice. 

**Describe alternatives you've considered**
Just having an SSH user for the CI jobs.

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by jonringer** [comment]
you should be able to do `git+ssh://`and that will use user credentials

**Comment by jkarni** [comment]
> you should be able to do `git+ssh://`and that will use user credentials

Yes, I’m aware. But on CI tokens are much easier than SSH keys, and the issue is that there’s no *single* protocol that works for both SSH and tokens, so you end up forcing either users or CI to use a suboptimal auth method.
